pubsub/src/client.rs

```diff
@@ -62,25 +102,37 @@ impl Client {
         Self::new(Default::default()).await
     }
 
-    /// new creates a Pub/Sub client.
+    /// new creates a Pub/Sub client. See [`ClientConfig`] for more information.
     pub async fn new(config: ClientConfig) -> Result<Self, Error> {
         let pool_size = config.pool_size.unwrap_or_default();
-        let environment = match std::env::var("PUBSUB_EMULATOR_HOST") {
-            Ok(host) => Environment::Emulator(host),
-            Err(_) => Environment::GoogleCloud(google_cloud_auth::project().await?),
+
+        let (project_id, environment) = match config.project {
+            ProjectOptions::Emulated(host) => {
+                let host = host.ok_or(Error::EmulatorHostNotSet)?;
+                (host.clone(), Environment::Emulator(host))
```

Comment:
In this implementation the `projectId` will be the `host`.

It would be better if the `projectId` can be set to an arbitrary value for a project or emulator that is separate from the GCP running this library.

How about setting `project` to `None` by default and changing it if necessary?

```rust
#[derive(Debug)]
pub enum ProjectOptions {
    Emulated(String),
    Project(Option<Project>),
}

#[derive(Debug)]
pub struct ClientConfig {
    pub pool_size: Option<usize>,
    
    pub project_id: Option<String>,

    pub project: ProjectOptions,

    pub endpoint: String,
}

impl Default for ClientConfig {
    fn default() -> Self {
        let host = std::env::var("PUBSUB_EMULATOR_HOST").ok();
        let options = match host {
            Some(host) => ProjectOptions::Emulated(host),
            None => ProjectOptions::Project(None)
        };
        Self {
            pool_size: Some(4),
            project: options,
            project_id: None,
            endpoint: PUBSUB.to_string(),
        }
    }
}

impl ClientConfig {
    pub fn project(&mut self, project: Project) {
        if let ProjectOptions::Project(_) = config.project {
            self.project = ProjectOptions::Project(Some(project))
        }
    }
}

impl Client {
    /// default creates a default Pub/Sub client.
    pub async fn default() -> Result<Self, Error> {
        Self::new(Default::default()).await
    }

    /// new creates a Pub/Sub client.
    pub async fn new(config: ClientConfig) -> Result<Self, Error> {
        let pool_size = config.pool_size.unwrap_or_default();
        let environment = match config.project {
            ProjectOptions::Emulated(host) => Environment::Emulator(host),
            ProjectOptions::Project(project) => {
                let project = match project {
                    Some(project) => project,
                    None => google_cloud_auth::project().await?
                };
                Environment::GoogleCloud(project)
            },
        };
        let pubc =
            PublisherClient::new(ConnectionManager::new(pool_size, &environment, config.endpoint.as_str()).await?);
        let subc =
            SubscriberClient::new(ConnectionManager::new(pool_size, &environment, config.endpoint.as_str()).await?);

        let project_id = match config.project_id {
            Some(project_id) => project_id,
            None => match environment {
                Environment::GoogleCloud(project) => project.project_id().ok_or(Error::ProjectIdNotFound)?.to_string(),
                Environment::Emulator(_) => "local-project".to_string(),
            },
        };
        Ok(Self { project_id, pubc, subc })
    }

```

The usage of this code looks like this:
```rust
let  mut config = ClientConfig::default();

// If you want to specify a different GCP project
// config.project_id = Some("other gcp project id");

// If you want to reuse an already created `Project` without executing `google_cloud_auth::project()` in `client::new`
// config.project(project);

let client = Client::new(config).await?;


```

Reply:
Thanks for the review. I agree with your changes and have them implemented. My only concern is it looks like `Environment::Emulator(String)` the host address is not being used, if it's always going to be local-project can we just remove the need for the `String`?

Reply:
I'm all good with these changes. I have them implemented let me know if there are any other concerns.